File: scratch/extracted/bellman-ford/optimize.java


public static class Solution {
    /**
     * Finds the cheapest price from src to dst with at most k stops using Bellman-Ford algorithm.
     *
     * @param n       The number of cities.
     * @param flights The list of flights, where flights[i] = [fromi, toi, pricei].
     * @param src     The source city.
     * @param dst     The destination city.
     * @param k       The maximum number of stops allowed.
     * @return The cheapest price from src to dst with at most k stops, or -1 if no such route exists.
     */
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        // Initialize an array to store the minimum prices to each city from the source.
        // Initially, set all prices to infinity except for the source city, which is 0.
        int[] prices = new int[n];
        Arrays.fill(prices, Integer.MAX_VALUE);
        prices[src] = 0;

        // Iterate k+1 times, where k is the maximum number of stops allowed.
        // Each iteration relaxes the edges, updating the minimum prices.
        for (int i = 0; i <= k; i++) {
            // Create a temporary array to store the prices for this iteration.
            // This prevents updates from affecting other calculations within the same iteration.
            int[] tmpPrices = Arrays.copyOf(prices, n);

            // Iterate through each flight in the flights array.
            for (int[] flight : flights) {
                int s = flight[0]; // Source city of the flight.
                int d = flight[1]; // Destination city of the flight.
                int p = flight[2]; // Price of the flight.

                // If the price to reach the source city 's' is still infinity,
                // it means it's not reachable, so we skip this flight.
                if (prices[s] == Integer.MAX_VALUE) continue;

                // If the price to reach the destination city 'd' through city 's'
                // is cheaper than the current minimum price to reach 'd',
                // update the temporary prices array.
                if (prices[s] + p < tmpPrices[d]) {
                    tmpPrices[d] = prices[s] + p;
                }
            }

            // Update the prices array with the values from the temporary array.
            // This ensures that the prices are updated after each iteration.
            prices = Arrays.copyOf(tmpPrices, n);
        }

        // After k+1 iterations, the prices array contains the minimum prices to each city
        // with at most k stops. If the price to reach the destination city 'dst' is still infinity,
        // it means it's not reachable within k stops, so we return -1.
        // Otherwise, we return the minimum price to reach 'dst'.
        return prices[dst] == Integer.MAX_VALUE ? -1 : prices[dst];
    }
}

File: scratch/extracted/bellman-ford/optimize.py
def findCheapestPrice(n: int, flights: list[list[int]], src: int, dst: int, k: int) -> int:
    # Initialize prices array with infinity for all cities except the source.
    prices = [float('inf')] * n
    prices[src] = 0

    # Iterate k+1 times, where k is the maximum number of stops allowed.
    # Each iteration represents considering paths with up to 'i' stops.
    for i in range(k + 1):
        # Create a temporary array to store the prices for the current iteration.
        # This prevents updates from prematurely affecting the calculations.
        tmp_prices = prices[:]

        # Iterate through all the flights.
        for s, d, p in flights:
            # If the price to reach the source city 's' is infinity,
            # it means we cannot reach this city yet, so skip this flight.
            if prices[s] == float('inf'):
                continue

            # If taking the current flight [s, d, p] improves the minimum price to reach
            # the destination city 'd', update the temporary price.
            if prices[s] + p < tmp_prices[d]:
                tmp_prices[d] = prices[s] + p

        # Update the prices array with the temporary prices after each iteration.
        # This ensures that we're considering the best prices with up to 'i' stops.
        prices = tmp_prices[:]

    # After k+1 iterations, the prices array contains the minimum prices to reach each city.
    # Return the price to reach the destination city 'dst' if it's not infinity.
    # Otherwise, return -1 to indicate that the destination is not reachable within k stops.
    return prices[dst] if prices[dst] != float('inf') else -1